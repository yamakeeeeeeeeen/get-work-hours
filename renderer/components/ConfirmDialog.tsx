import { FC, memo } from 'react';
import { Box, Button, Dialog, Typography } from '@material-ui/core';

type Props = {
  isOpen: boolean;
  setClose: () => void;
  body: string;
  callback: () => void;
};

export const ConfirmDialog: FC<Props> = memo(({ isOpen, setClose, body, callback }) => {
  return (
    <>
      <Dialog open={isOpen}>
        <Typography variant="h4">{body}</Typography>
        <Box display="flex">
          <Button variant="contained" color="primary" onClick={callback}>
            Yes
          </Button>
          <Button variant="contained" color="secondary" onClick={setClose}>
            No
          </Button>
        </Box>
      </Dialog>
    </>
  );
});

ConfirmDialog.displayName = 'ConfirmDialog';
