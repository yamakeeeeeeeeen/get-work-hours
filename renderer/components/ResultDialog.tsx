import { FC, memo } from 'react';
import { Box, Button, Dialog, Typography } from '@material-ui/core';

type Props = {
  isOpen: boolean;
  setClose: () => void;
  result: { breakTime: string; uptime: string };
};

export const ResultDialog: FC<Props> = memo(({ isOpen, setClose, result }) => {
  return (
    <>
      <Dialog open={isOpen}>
        <Typography variant="h4">{result.uptime}</Typography>
        <Typography variant="h4">{result.breakTime}</Typography>
        <Box display="flex">
          <Button variant="contained" color="secondary" onClick={setClose}>
            CLOSE
          </Button>
        </Box>
      </Dialog>
    </>
  );
});

ResultDialog.displayName = 'ResultDialog';
