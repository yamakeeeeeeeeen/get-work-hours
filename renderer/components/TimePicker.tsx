import { createStyles, makeStyles, TextField, Theme } from '@material-ui/core';
import { FC } from 'react';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    textField: {
      marginLeft: theme.spacing(1),
      marginRight: theme.spacing(1),
      width: 120,
    },
  }),
);

type Props = {
  label: string;
};

export const TimePicker: FC<Props> = ({ label }) => {
  const classes = useStyles();

  return (
    <TextField
      label={label}
      type="time"
      defaultValue="12:00"
      variant="outlined"
      color="primary"
      className={classes.textField}
      InputLabelProps={{
        shrink: true,
      }}
    />
  );
};
