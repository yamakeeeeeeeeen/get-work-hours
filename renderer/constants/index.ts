export const FRACTIONS_OF_AN_HOUR = 60;
