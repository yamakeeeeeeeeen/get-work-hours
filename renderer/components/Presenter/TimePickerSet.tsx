import { FC, memo } from 'react';
import { Box, Button, Typography } from '@material-ui/core';
import { TimePicker } from '~/components/TimePicker';
import { Register } from '~/@types/ReactHookForm';
import { ConfirmDialog } from '~/components/ConfirmDialog';

type ComponentProps = {
  index: number;
  value: {
    start: string;
    end: string;
  };
  isDisabled: boolean;
  isOpen: boolean;
  setOpen: () => void;
  setClose: () => void;
  register: Register;
  handleRemove: () => void;
};

export const Component: FC<ComponentProps> = memo(
  ({ index, value, isDisabled, isOpen, setOpen, setClose, register, handleRemove }) => {
    return (
      <>
        <Box display="flex" alignItems="center" mb={3}>
          <Typography>着席{index + 1}回目</Typography>
          <TimePicker
            name={`workTimes[${index}].start`}
            label="着席"
            defaultValue={value.start}
            inputRef={register()}
          />
          <Typography>~</Typography>
          <TimePicker name={`workTimes[${index}].end`} label="離席" defaultValue={value.end} inputRef={register()} />
          <Button variant="contained" color="secondary" disabled={isDisabled} onClick={setOpen}>
            削除
          </Button>
        </Box>
        <ConfirmDialog isOpen={isOpen} setClose={setClose} body="削除してもよろしいですか？" callback={handleRemove} />
      </>
    );
  },
);

Component.displayName = 'Component';
