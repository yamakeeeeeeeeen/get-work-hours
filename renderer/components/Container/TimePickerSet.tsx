import { FC, memo, useCallback, useMemo } from 'react';
import { useFormContext, UseFieldArrayMethods } from 'react-hook-form';
import { Component } from '~/components/Presenter/TimePickerSet';
import useBooleanState from '~/hooks/useBooleanState';

type Props = {
  index: number;
  value: {
    start: string;
    end: string;
  };
  fields: UseFieldArrayMethods['fields'];
  remove: UseFieldArrayMethods['remove'];
};

export const TimePickerSet: FC<Props> = memo(({ fields, remove, ...props }) => {
  const { errors, getValues, register } = useFormContext();

  const isDisabled = useMemo<boolean>(() => fields.length === 1, [fields.length]);
  const [isOpen, setOpen, setClose] = useBooleanState(false);

  const handleRemove = useCallback(() => {
    remove(props.index);
    setClose();
  }, [props.index, remove, setClose]);

  return (
    <Component
      {...props}
      isDisabled={isDisabled}
      isOpen={isOpen}
      setOpen={setOpen}
      setClose={setClose}
      errors={errors}
      getValues={getValues}
      register={register}
      handleRemove={handleRemove}
    />
  );
});

TimePickerSet.displayName = 'TimePickerSet';
