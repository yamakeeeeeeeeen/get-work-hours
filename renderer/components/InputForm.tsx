import { FC, memo } from 'react';
import { Box, Button, createStyles, makeStyles, Theme, Typography } from '@material-ui/core';
import { Props } from '~/containers/InputForm';
import { TimePickerSet } from '~/containers/TimePickerSet';
import { ResultDialog } from '~/components/ResultDialog';
import { Inputs } from '~/pages';
import { UseFormMethods } from 'react-hook-form';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    infoText: {
      marginBottom: theme.spacing(3),
    },
  }),
);

export type ComponentProps = Omit<Props, 'append'> & {
  result: { breakTime: string; uptime: string };
  isDisabled: boolean;
  isOpen: boolean;
  setClose: () => void;
  handleAppend: () => void;
  handleConfirm: (val: Inputs) => void;
  handleReset: () => void;
  handleSubmit: UseFormMethods['handleSubmit'];
};

export const Component: FC<ComponentProps> = memo(
  ({ result, isDisabled, isOpen, setClose, handleAppend, handleConfirm, handleReset, handleSubmit, ...props }) => {
    const classes = useStyles();

    return (
      <>
        <Box maxWidth={500} mx="auto">
          <Typography className={classes.infoText}>Please enter the hours you workedüë®‚Äçüíª</Typography>
          {props.fields.map((item, index) => {
            if (item.start && item.end) {
              const value = {
                start: item.start,
                end: item.end,
              };
              return <TimePickerSet key={item.id} index={index} value={value} {...props} />;
            }
            return null;
          })}
          <Box display="flex" justifyContent="space-between">
            <Button variant="contained" color="primary" onClick={handleAppend}>
              APPEND
            </Button>
            <Button variant="contained" color="primary" onClick={handleSubmit(handleConfirm)}>
              CONFIRM
            </Button>
            <Button variant="contained" color="secondary" disabled={!isDisabled} onClick={handleReset}>
              RESET
            </Button>
          </Box>
        </Box>
        <ResultDialog isOpen={isOpen} setClose={setClose} result={result} />
      </>
    );
  },
);

Component.displayName = 'Component';
