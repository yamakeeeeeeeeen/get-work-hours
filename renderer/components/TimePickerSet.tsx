import { FC, memo } from 'react';
import { UseFormMethods } from 'react-hook-form';
import { Box, Button, Typography } from '@material-ui/core';
import { Props } from '~/containers/TimePickerSet';
import { TimePicker } from '~/components/TimePicker';
import { ConfirmDialog } from '~/components/ConfirmDialog';
import { getValidation } from '~/utils/getValidation';
import { ORDINAL_NUMBERS } from '~/constants';

type ComponentProps = Omit<Props, 'fields' | 'remove'> & {
  isDisabled: boolean;
  isOpen: boolean;
  setOpen: () => void;
  setClose: () => void;
  errors: UseFormMethods['errors'];
  getValues: UseFormMethods['getValues'];
  register: UseFormMethods['register'];
  handleRemove: () => void;
};

export const Component: FC<ComponentProps> = memo(
  ({ index, value, isDisabled, isOpen, setOpen, setClose, errors, getValues, register, handleRemove }) => {
    const { biggerThanThePrevious, formatHHMM } = getValidation();

    return (
      <>
        <Box display="flex" alignItems="center" mb={3}>
          <Typography>{ORDINAL_NUMBERS[index + 1]}</Typography>
          <TimePicker
            name={`workTimes[${index}].start`}
            label="start"
            defaultValue={value.start}
            inputRef={register({
              validate: {
                ...formatHHMM(),
                ...biggerThanThePrevious('start', index, getValues),
              },
            })}
            error={!!errors?.workTimes?.[index]?.start}
            helperText={errors?.workTimes?.[index]?.start && errors?.workTimes?.[index]?.start?.message}
          />
          <Typography>~</Typography>
          <TimePicker
            name={`workTimes[${index}].end`}
            label="end"
            defaultValue={value.end}
            inputRef={register({
              validate: {
                ...formatHHMM(),
                ...biggerThanThePrevious('end', index, getValues),
              },
            })}
            error={!!errors?.workTimes?.[index]?.end}
            helperText={errors?.workTimes?.[index]?.end && errors?.workTimes?.[index]?.end?.message}
          />
          <Button variant="contained" color="secondary" disabled={isDisabled} onClick={setOpen}>
            DELETE
          </Button>
        </Box>
        <ConfirmDialog isOpen={isOpen} setClose={setClose} body="Can I delete it?" callback={handleRemove} />
      </>
    );
  },
);

Component.displayName = 'Component';
