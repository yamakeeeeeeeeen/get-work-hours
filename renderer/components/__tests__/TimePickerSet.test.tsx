import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Component as TimePickerSet, ComponentProps as Props } from '~/components/TimePickerSet';
import { ORDINAL_NUMBERS } from '~/constants';

const setOpen = jest.fn();
const setClose = jest.fn();
const errors = jest.fn();
const getValues = jest.fn();
const register = jest.fn();
const handleRemove = jest.fn();
const props: Props = {
  index: 0,
  value: { start: '09:00', end: '18:00' },
  isDisabled: true,
  isOpen: true,
  setOpen,
  setClose, // propsで渡すのみ
  errors,
  getValues,
  register,
  handleRemove, // propsで渡すのみ
};
const wrapper = shallow(<TimePickerSet {...props} />);

describe('<TimePickerSet />', () => {
  it('snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('times should be displayed as "first"', () => {
    expect(wrapper.find('.times').text()).toBe(ORDINAL_NUMBERS[props.index + 1]);
  });

  it('setOpen should be called', () => {
    wrapper.find('.delete-btn').simulate('click');
    expect(setOpen).toHaveBeenCalled();
  });
});
