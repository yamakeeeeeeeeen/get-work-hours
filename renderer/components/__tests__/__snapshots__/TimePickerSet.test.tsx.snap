// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimePickerSet /> snapshot 1`] = `
<Fragment>
  <Styled(MuiBox)
    alignItems="center"
    display="flex"
    justifyContent="space-between"
    mb={3}
  >
    <Styled(MuiBox)
      width={75}
    >
      <WithStyles(ForwardRef(Typography))
        className="times"
      >
        first
      </WithStyles(ForwardRef(Typography))>
    </Styled(MuiBox)>
    <Styled(MuiBox)
      alignItems="center"
      display="flex"
    >
      <TimePicker
        defaultValue="09:00"
        error={false}
        label="start"
        name="workTimes[0].start"
      />
      <Styled(MuiBox)
        mx={2}
      >
        <WithStyles(ForwardRef(Typography))>
          ~
        </WithStyles(ForwardRef(Typography))>
      </Styled(MuiBox)>
      <TimePicker
        defaultValue="18:00"
        error={false}
        label="end"
        name="workTimes[0].end"
      />
    </Styled(MuiBox)>
    <WithStyles(ForwardRef(Button))
      className="delete-btn"
      color="secondary"
      disabled={true}
      onClick={[MockFunction]}
      variant="contained"
    >
      DELETE
    </WithStyles(ForwardRef(Button))>
  </Styled(MuiBox)>
  <ConfirmDialog
    body="Can I delete it?"
    callback={[MockFunction]}
    isOpen={true}
    setClose={[MockFunction]}
  />
</Fragment>
`;
